x-pack/plugins/apm/public/hooks/useFetcher.tsx

```diff
@@ -27,21 +27,28 @@ export function useFetcher<Opts, Response>(
     error?: Error;
   }>({});
 
-  const useEffectKey = [...Object.keys(options), ...Object.values(options)];
+  const useEffectKey = [...Object.keys(fnOptions), ...Object.values(fnOptions)];
 
   useEffect(() => {
     let didCancel = false;
+    let didFinish = false;
 
-    dispatchStatus({ name: fn.name, isLoading: true });
-
-    setResult({
-      data: result.data, // preserve data from previous state while loading next state
-      status: FETCH_STATUS.LOADING,
-      error: undefined
+    // only apply the loading indicator if the promise did not resolve immediately
+    // the promise will resolve immediately if the value was found in cache
+    requestAnimationFrame(() => {
+      if (!didFinish && !didCancel) {
+        dispatchStatus({ name: fn.name, isLoading: true });
+        setResult({
+          data: result.data, // preserve data from previous state while loading next state
+          status: FETCH_STATUS.LOADING,
+          error: undefined
+        });
+      }
     });
 
-    fn(options)
+    fn(fnOptions)
       .then(data => {
```

Comment:
`useEffect` does not support async/await. I could get around it by calling an async function from `useEffect` but this seemed clearer. I'm okay with moving it out it you prefer.

Reply:
Wow that's weird. I think it's fine then but can you add a comment there like:
```
// useEffect doesn't accept async functions
// https://github.com/facebook/react/issues/14326#issuecomment-441680293
```
otherwise I know every person who visits this file is going to try to rewrite it to async/await and then find the console error :D

Reply:
I used the async example from Abramov.